compiler/vm/vmdef.nim

```diff
@@ -90,27 +90,138 @@ type
-    slotFixedLet,     # slot is used for a fixed param/let
-    slotTempUnknown,  # slot but type unknown (argument of proc call)
-    slotTempInt,      # some temporary int
-    slotTempFloat,    # some temporary float
-    slotTempStr,      # some temporary string
-    slotTempComplex,  # some complex temporary (s.node field is used)
-    slotTempPerm      # slot is temporary but permanent (hack)
-
-
-  TFullReg* = object  # with a custom mark proc, we could use the same
-                      # data representation as LuaJit (tagged NaNs).
-    case kind*: TRegisterKind
-    of rkNone: nil
-    of rkInt: intVal*: BiggestInt
-    of rkFloat: floatVal*: BiggestFloat
-    of rkNode: node*: PNode
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    slotEmpty,        ## slot is unused
+    slotFixedVar,     ## slot is used for a fixed var/result (requires copy then)
+    slotFixedLet,     ## slot is used for a fixed param/let
+    slotTempUnknown,  ## slot but type unknown (argument of proc call)
+    slotTempInt,      ## some temporary int
+    slotTempFloat,    ## some temporary float
+    slotTempStr,      ## some temporary string
+    slotTempComplex,  ## some complex temporary (s.node field is used)
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    akInt
+    akFloat
+    akString # TODO: maybe even collapse akString into akOrdinalArray?
+    akOrdinalArray # Space efficient storing of float/int arrays
+    akArray # Used for both seqs and fixed size arrays by vmgen
+    akSet
+    akRef
+    akPtr
+    akPointer ## A raw pointer points to inside akString's string or akOrdinalArray/akArray's rawArrayVal
+    akObject
+    akPNode
+    akCallable
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
+    of ckTemplate:
+      templateSym*: PSym
+
+
+  OrdinalArrayKind* = enum
+    oakUnsigned
+    oakSigned
+    oakFloat32
+    oakFloat
+
+  VmObjectAtom* = object
+    case kind*: AtomKind
+    of akInt:
+      # XXX: is it okay to use this for unsigned integer too?
+      intVal*: BiggestInt
+    of akFloat:
+      floatVal*: BiggestFloat
+    of akString:
+      strVal*: string
+    of akOrdinalArray:
+      rawArrayVal*: seq[byte]
+      elemSize*: int8 # 1, 2, 4 or 8
+      oaKind*: OrdinalArrayKind # TODO: encode `oaKind` and `elemTyp` into a single int8
+    of akArray:
+      arrayElements*: seq[VmObjectAtom]
+      elemTyp*: PType
```

Comment:
I think it might be better to merge arrays, objects and strings into one unified representation via a `seq[byte]`. Nim objects and friends can't have their type mutated at runtime, so there shouldn't be a need to store them in the array itself.

An idea that I have is to have a constant array designed to store type info in `TCtx`, then change your `kind: AtomKind` into something like `kind: TypeId`, which will be the index to that array describing the type of the atom, with the positions corresponding to `AtomKind` reserved.

Reply:
That sounds really similar to what I'm attempting right now with the flat object idea I mentioned.

Using @saem 's proposed terminology (`location`, `handle`):  
- A `seq[byte]` stores a sequence of `location`s
- A `location` is either an atom (int, float, seq, etc.) or a "container"
- A "container" is either homogeneous (`array`) or heterogeneous (`object`) and contains either atoms or other "containers"
- A `handle` is a reference to a `location`

To interpret the `location` referenced by a `handle` you then need the corresponding type info (`TypeInfo`) which is stored alongside the `handle` as a handle itself (`typeHandle`). The `TypeInfo` stores whether a `location` is an atom or a "container" and also the size in bytes of the `location`. 
Additionally:
- For homogeneous "containers", `TypeInfo` stores:
  - element count (`array` size is part of the type info)
  - element stride (number of bytes between two elements)
  - element type (handle to a `TypeInfo`)
- For heterogeneous "containers", `TypeInfo` stores a sequence of field information where each stores:
  - field offset (number of bytes from the start of the "container")
  - field type (handle to a `TypeInfo`)
- For `seq` and `set` atoms, `TypeInfo` stores the respective element types.
- For `ptr` and `ref` atoms, `TypeInfo` stores the respective target types

I'm almost done with a first prototype in my local repository

Reply:
I rewrote this comment deleting tons of content over and over again, so I think I'm missing some things here. This is the basis for my thinking, a quote from the manual:

> code specifies a computation that acts on a memory consisting of components called locations. A variable is basically a name for a location. Each variable and location is of a certain type. The variable's type is called static type, the location's type is called dynamic type. If the static type is not the same as the dynamic type, it is a super-type or subtype of the dynamic type.

So the above sounds pretty close. For the purposes of the VM, we need to decide the level of operation we support:
* once we're in pointer land, then the memory model needs to behave in a flat fashion, a la C, this is very baked in; can't change it, yet
* so if we're in pointer land then arrays, seq, and objects need to have their parts stored contiguously
  * so `seq[byte]` works at least for that
  * we lose "free" gc; but the complexity isn't worth it

From the quote you'll notice that there is the dynamic memory model (locations) and the static memory model (handles). For the purposes of the VM. If we decide that the static/handle memory model is the floor of abstraction, then we're saying that location is a higher level concepts in VM land. This is easier to implement, but I also think we won't be able to run a bunch of key code if we go this route.

I unfortunately don't have a particularly conclusive remark right now, so I'm going to go look at the VM again and remind myself of how it uses `PNode`s. I suspect giving the VM its own explict "heap" (a `seq[PNode]`) and could have it work exclusively on those; outside of copying data into and then copying results out of for compile-time evaluation is a good firs step. The problem remains of AST manipulations for meta-programming, but that's a completely different use of `PNode` IMO.